api/src/main/java/org/apache/iceberg/transforms/ProjectionUtil.java

```diff
@@ -185,20 +145,117 @@ private ProjectionUtil() {}
+        if (boundary.equals(transform.apply(boundary))) {
+          // the partition only contains this single value therefore would be same as the predicate
+          return predicate(Expression.Operation.GT, name, transform.apply(boundary));
+        } else {
+          return predicate(Expression.Operation.GT_EQ, name, transform.apply(boundary));
+        }
       case GT_EQ:
         return predicate(Expression.Operation.GT_EQ, name, transform.apply(boundary));
       case EQ:
         return predicate(Expression.Operation.EQ, name, transform.apply(boundary));
-//        case IN: // TODO
-//          return Expressions.predicate(Operation.IN, name, transform.apply(boundary));
+      case NOT_EQ:
+        // If the boundary is same as the partition value, it implies
+        // that the truncate width is more than the size of the array
+        if (boundary.equals(transform.apply(boundary))) {
+          // the partition only contains this single value therefore would be same as the predicate
+          return predicate(Expression.Operation.NOT_EQ, name, transform.apply(boundary));
+        } else {
+          return null;
+        }
+      default:
+        return null;
+    }
+  }
+
+  static <S, T> UnboundPredicate<T> truncateArrayStrict(
+      String name, BoundPredicate<S> pred, Transform<S, T> transform) {
+    S boundary = pred.literal().value();
+    switch (pred.op()) {
+      case LT:
+        return predicate(Expression.Operation.LT, name, transform.apply(boundary));
+      case LT_EQ:
+        // If the boundary is same as the partition value, it implies
+        // that the truncate width is more than the size of the array
+        if (boundary.equals(transform.apply(boundary))) {
+          // the partition only contains this single value therefore would be same as the predicate
+          return predicate(Expression.Operation.LT_EQ, name, transform.apply(boundary));
+        } else {
+          return predicate(Expression.Operation.LT, name, transform.apply(boundary));
+        }
+      case GT:
+        return predicate(Expression.Operation.GT, name, transform.apply(boundary));
+      case GT_EQ:
+        // If the boundary is same as the partition value, it implies
+        // that the truncate width is more than the size of the array
+        if (boundary.equals(transform.apply(boundary))) {
+          // the partition only contains this single value therefore would be same as the predicate
+          return predicate(Expression.Operation.GT_EQ, name, transform.apply(boundary));
+        } else {
+          return predicate(Expression.Operation.GT, name, transform.apply(boundary));
+        }
+      case NOT_EQ:
+        return predicate(Expression.Operation.NOT_EQ, name, transform.apply(boundary));
+      case EQ:
+        // If the boundary is same as the partition value, it implies
+        // that the truncate width is more than the size of the array
+        if (boundary.equals(transform.apply(boundary))) {
```

Comment:
I agree with this when the array is shorter than the boundary. So "ab" with truncate length 3 is always "ab" and nothing else can be "ab". But "abc" with truncate length 3 would also contain "abcd", which is not equal to "abc". So I think this can only be the case if the length of the string (in code points) is less than the truncate length.

I'm not sure that handling this case is worth it after all. Maybe this should just return null in all cases.

Reply:
Ah, I see what you mean. It's not guaranteed to be a partition with the same value and can contain others. Will revert this change 👍